File: app/src/main/java/com/example/pearvideoclient/home/fragment/CityFragment.java
package com.example.pearvideoclient.home.fragment;

import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.pearvideoclient.CommonCallBack;
import com.example.pearvideoclient.Constants;
import com.example.pearvideoclient.R;
import com.example.pearvideoclient.entity.LocalContEntity;
import com.example.pearvideoclient.utils.ScreenUtils;
import com.scwang.smartrefresh.layout.api.RefreshLayout;

import java.util.ArrayList;
import java.util.List;

/**
 * @Description: java类作用描述
 * @Author: jiazhu
 * @Date: 2019-08-21 15:03
 * @ClassName: CityFragment
 */
public class CityFragment extends Fragment {

    private RefreshLayout mRefreshLayout;
    private RecyclerView mRvCityList;

    private CityAdapter mCityAdapter;

    private CommonCallBack<Integer, Void> mCityFragmentCallBack;
    private Context mContext;

    public void setCityFragmentCallBack(CommonCallBack<Integer, Void> cityFragmentCallBack) {
        mCityFragmentCallBack = cityFragmentCallBack;
    }

    public static CityFragment newInstance() {
        Bundle bundle = new Bundle();
        CityFragment fragment = new CityFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_city, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initView(view);
        initData();
    }

    private void initView(View view) {
        mRefreshLayout = view.findViewById(R.id.refresh_layout);
        mRvCityList = view.findViewById(R.id.rv_city_list);
    }

    private void initData() {
        mContext = getActivity();
        mRvCityList.setLayoutManager(new LinearLayoutManager(mContext, LinearLayoutManager.VERTICAL, false));
        mCityAdapter = new CityAdapter(new ArrayList<>());
        mRvCityList.setAdapter(mCityAdapter);

        mRefreshLayout.setOnLoadMoreListener(refreshLayout -> {
            if (mCityFragmentCallBack != null) {
                mCityFragmentCallBack.todo(Constants.LOAD_MORE);
            }
        });
        mRefreshLayout.setOnRefreshListener(refreshLayout -> {
        });
        if (mCityFragmentCallBack != null) {
            mCityFragmentCallBack.todo(Constants.LOAD_REFRESH);
        }

        addHeaderView();
    }

    /**
     * 添加headerView
     */
    private void addHeaderView() {
        View headerView = getLayoutInflater().inflate(R.layout.item_city_header_layout, null);
        headerView.setLayoutParams(new DrawerLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ScreenUtils.dip2px(mContext, 48)));
        headerView.setOnClickListener(v -> {
            if (mCityFragmentCallBack != null) {
                mCityFragmentCallBack.todo(Constants.HEADER_CLICK);
            }
        });
        mCityAdapter.setHeaderView(headerView);
    }

    public void addData(List<LocalContEntity> data) {
        mCityAdapter.addData(data);
    }

    public void replaceData(List<LocalContEntity> data) {
        mCityAdapter.replaceData(data);
    }

    public void loadMoreFinish(boolean isSuccess) {
        mRefreshLayout.finishLoadMore(isSuccess);
    }

    public void refreshFinish(boolean isSuccess) {
        mRefreshLayout.finishRefresh(isSuccess);
    }
}

File: app/src/main/java/com/example/pearvideoclient/home/fragment/CityFragment.kt
package com.example.pearvideoclient.home.fragment

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.apublic.CommonCallBack
import com.example.pearvideoclient.Constants
import com.example.pearvideoclient.R
import com.example.pearvideoclient.entity.LocalContEntity
import com.example.pearvideoclient.utils.ScreenUtils
import kotlinx.android.synthetic.main.fragment_city.*
import java.util.*

/**
 * @Description: java类作用描述
 * @Author: jiazhu
 * @Date: 2019-08-21 15:03
 * @ClassName: CityFragment
 */
class CityFragment : Fragment() {

    private var mCityAdapter: CityAdapter? = null

    private var mCityFragmentCallBack: CommonCallBack<Int, Unit>? = null
    private var mContext: Context? = null

    fun setCityFragmentCallBack(cityFragmentCallBack: CommonCallBack<Int, Unit>) {
        mCityFragmentCallBack = cityFragmentCallBack
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_city, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initData()
    }

    private fun initData() {
        mContext = activity
        rv_city_list.layoutManager = LinearLayoutManager(mContext, LinearLayoutManager.VERTICAL, false)
        mCityAdapter = CityAdapter(ArrayList())
        rv_city_list.adapter = mCityAdapter

        refresh_layout.setOnLoadMoreListener {
            if (mCityFragmentCallBack != null) {
                mCityFragmentCallBack!!.todo(Constants.LOAD_MORE)
            }
        }
        refresh_layout.setOnRefreshListener {
            if (mCityFragmentCallBack != null) {
                mCityFragmentCallBack!!.todo(Constants.LOAD_REFRESH)
            }
        }


        addHeaderView()
    }

    /**
     * 添加headerView
     */
    private fun addHeaderView() {
        val headerView = layoutInflater.inflate(R.layout.item_city_header_layout, null)
        headerView.layoutParams = DrawerLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ScreenUtils.dip2px(mContext!!, 48f))
        headerView.setOnClickListener {
            if (mCityFragmentCallBack != null) {
                mCityFragmentCallBack!!.todo(Constants.HEADER_CLICK)
            }
        }
        mCityAdapter!!.setHeaderView(headerView)
    }

    fun addData(data: List<LocalContEntity>) {
        mCityAdapter!!.addData(data)
    }

    fun replaceData(data: List<LocalContEntity>) {
        mCityAdapter!!.replaceData(data)
    }

    fun loadMoreFinish(isSuccess: Boolean) {
        refresh_layout.finishLoadMore(isSuccess)
    }

    fun refreshFinish(isSuccess: Boolean) {
        refresh_layout.finishRefresh(isSuccess)
    }

    companion object {

        fun newInstance(): CityFragment {
            val bundle = Bundle()
            val fragment = CityFragment()
            fragment.arguments = bundle
            return fragment
        }
    }
}
